MetaMorpheus/EngineLayer/Util/IsobaricMassTag.cs

```diff
@@ -0,0 +1,227 @@
+            }
+
+            // Map the enum to the modification name pattern
+            string modificationPattern = GetModificationPattern((IsobaricMassTagType)tagType);
+
+            // Find the corresponding modification in GlobalVariables.AllModsKnown
+            Modification isobaricModification = GlobalVariables.AllModsKnown
+                .FirstOrDefault(m => m.ModificationType == "Multiplex Label"
+                                     && m.IdWithMotif.Contains(modificationPattern));
+
+            // Verify that the modification exists and has diagnostic ions
+            if (isobaricModification == null
+                || !isobaricModification.DiagnosticIons.TryGetValue(DissociationType.HCD, out var reporterIonMasses) 
+                || reporterIonMasses.IsNullOrEmpty()) 
+            {
+                return null;
+            }
+
+            // Extract the reporter ion m/z values from the first dissociation type's diagnostic ions
+            // and convert them to m/z values (charge state 1)
+            double[] reporterIonMzs = reporterIonMasses
+                .Select(mass => mass.ToMz(1))
+                .OrderBy(mz => mz)
+                .ToArray();
+
+            return new IsobaricMassTag
+            {
+                TagType = (IsobaricMassTagType)tagType,
+                ReporterIonMzs = reporterIonMzs,
+                Tolerance = new AbsoluteTolerance(AbsoluteToleranceValue)
+            };
+        }
+
+        /// <summary>
+        /// Takes in a spectrum, searches for the reporter ions within the specified ppm tolerance,
+        /// and returns their intensities. If a reporter ion is not found, its intensity will be zero.
+        /// </summary>
+        public double[]? GetReporterIonIntensities(MzSpectrum spectrum)
+        {
+            if (spectrum == null || spectrum.Size < 1) return null;
+            double[] reporterIonIntensities = new double[ReporterIonMzs.Length];
+            if (spectrum.XArray[0] > Tolerance.GetMaximumValue(ReporterIonMzs[^1]))
+            {
+                return reporterIonIntensities; // If the spectrum starts after the last reporter ion, return all zeros.
+            }
+
+            // Two-pointer search. The for loop iterates through the theoretical ions mzs (theoreticalIonIndex),
+            // and the while loop advances through the spectrum peaks (spectrumIndex) until we find a match or surpass the theoretical mz.
+            // If there are multiple peaks that match the theoretical mz within tolerance, we take the one with the highest intensity.
+            int spectrumIndex = 0;
+            for (int theoreticalIonIndex = 0; theoreticalIonIndex < reporterIonIntensities.Length; theoreticalIonIndex++)
+            {
+                double minMz = Tolerance.GetMinimumValue(ReporterIonMzs[theoreticalIonIndex]);
+                double maxMz = Tolerance.GetMaximumValue(ReporterIonMzs[theoreticalIonIndex]);
+                double maxIntensity = 0;
```

Comment:
Optimization Opportunity: You are calculating the minMz and maxMz for each reporter ion by subtracting or adding 0.003 from each mz value. This is done for each reporter ion, for each spectrum. The tolerance values could be precalculated and accessed instead of calculated for each spectrum.